exercises/concept/take-a-number-deluxe/lib/take_a_number_deluxe/state.ex

```diff
@@ -0,0 +1,63 @@
+defmodule TakeANumberDeluxe.State do
+  defstruct min_number: 1, max_number: 999, queue: []
+
+  def new(min_number, max_number) do
+    if is_integer(min_number) and is_integer(max_number) and min_number < max_number do
+      {:ok, %__MODULE__{min_number: min_number, max_number: max_number}}
+    else
+      {:error, :invalid_configuration}
+    end
+  end
+
+  def queue_new_number(%__MODULE__{} = state) do
+    case find_next_available_number(state) do
+      {:ok, next_available_number} ->
+        {:ok, next_available_number, %{state | queue: [next_available_number | state.queue]}}
+
+      {:error, error} ->
+        {:error, error}
+    end
+  end
+
+  def serve_next_queued_number(%__MODULE__{} = state, priority_number) do
```

Comment:
For this function, the logic is a bit hard to follow from the code alone, I think it's coming from the fact that you are using a list to model a queue.
I have two ideas: either add some documentation, or use :queue. Something like (untested)

```elixir
cond do
  :queue.is_empty(state.queue) ->
    {:error, :empty_queue}

  is_nil(priority_number) ->
    {{:value, next_number}, new_queue} = :queue.out(state.queue)
    {:ok, next_number, %{state | queue: new_queue}}

  :queue.member(priority_number, state.queue) ->
    {:ok, priority_number, %{state | queue: :queue.delete(priority_number, state.queue)}}

  true ->
    {:error, :priority_number_not_found}
end    
```
Documentation is probably simpler :)

Reply:
I added specs and used `:queue`. In the instructions, I'll make it clear that those functions can be used as they are.